### PR TITLE
fix(diagnostics): fix navigation for diagnostics placed past end of line

### LINKED ISSUE #16349: Out of bounds diagnostics columns can prevent you from jumping past them using goto_next/goto_prev
### Neovim version (nvim -v)

NVIM v0.6.0-dev+588-geba317d7a

### Vim (not Nvim) behaves the same?

No

### Operating system/version

Arch Linux

### Terminal name/version

nvim-qt

### $TERM environment variable

nvim-qt

### Installation

Git

### How to reproduce the issue

Create a Lua file with the following contents:

```lua
vim.diagnostic.set(
  vim.api.nvim_create_namespace('test'),
  0,
  {
    { lnum = 1, col = 9000, severity = vim.diagnostic.severity.WARN, message = 'this is the first diagnostic' },
    { lnum = 2, col = 1, severity = vim.diagnostic.severity.WARN, message = 'this is the second diagnostic' },
  },
  {}
)
```

Save the file, then `:so` it. Next, try jumping to diagnostics using this:

```lua
vim.diagnostic.goto_next({ float = false, severity = { min = vim.diagnostic.severity.WARN } })
```

### Expected behavior

You are able to jump to every diagnostic

### Actual behavior

You get stuck on the first diagnostic, and can't jump over it. This is probably because the column being out of bounds results in NeoVim's `goto_next` finding it as the next diagnostic to jump to. Since it's out of bounds you can't actually jump to it, thus you get stuck.

### PR DESCRIPTION
**Pull request by gpanders**
Closes #16349.

These commits should not be squashed: the first contains a fix for clamping
line numbers that is relevant to this PR, but deserves its own commit message.

### CODE REVIEW

File: runtime/lua/vim/diagnostic.lua [hunk 1]
```diff
@@ -399,17 +397,56 @@ local function set_list(loclist, opts)
 end
 
 ---@private
---- To (slightly) improve performance, modifies diagnostics in place.
-local function clamp_line_numbers(bufnr, diagnostics)
-  local buf_line_count = vim.api.nvim_buf_line_count(bufnr)
-  if buf_line_count == 0 then
-    return
+local function get_diagnostics(bufnr, opts, clamp)
```
Comment (mjlbach) [inline runtime/lua/vim/diagnostic.lua:400]: Assuming for qflist we only need to clamp line number and not col number?

**Comment by mjlbach** [comment]
Did you check the overhead from going through the metamethod calls now? Would be nice just to benchmark (don't expect a regression).

File: runtime/lua/vim/diagnostic.lua (hunk 1 above)
Comment (gpanders) [reply runtime/lua/vim/diagnostic.lua:400]: We're not doing any clamping for qflist right now (though maybe we should). The only things that go through this code path are the navigation methods, `open_float()`, and `show()`.

Comment (mjlbach) [reply runtime/lua/vim/diagnostic.lua:400]: I thought the call chain was:

set_qflist -> set_list -> M.get -> get_diagnostics with clamp == false

Comment (mjlbach) [reply runtime/lua/vim/diagnostic.lua:400]: Ah sorry, I misunderstood your comment. I think we probably should clamp for qflist as well.

**Review by mjlbach — APPROVED** [review]

**Pull request merged**